contracts/brush/proc_macros/contract.rs

```diff
@@ -11,44 +12,150 @@ use proc_macro2::{
     TokenTree,
 };
 use fs2::FileExt;
+use crate::internal::*;
+use crate::storage_trait;
 
 pub(crate) fn generate(_attrs: TokenStream, ink_module: TokenStream) -> TokenStream {
     let input: TokenStream2 = ink_module.into();
     let attrs: TokenStream2 = _attrs.into();
     let mut module = syn::parse2::<syn::ItemMod>(input.clone()).expect("Can't parse contract module");
-    let (braces, items) = match module.content {
+    let (braces, mut items) = match module.content {
         Some((brace, items)) => (brace, items),
         None => {
             panic!(
                 "{}", "out-of-line ink! modules are not supported, use `#[ink::contract] mod name {{ ... }}`",
             )
         }
     };
+
+    // First we need to consume all traits and update metadata file.
+    // After we can consume all other stuff.
+    items = consume_traits(items);
+
     let locked_file = internal::get_locked_file();
-    let metadata = internal::load_metadata(&locked_file);
+    let metadata = internal::Metadata::load(&locked_file);
     locked_file.unlock().expect("Can't remove exclusive lock in extract_fields_and_methods");
 
-    let mut new_items: Vec<syn::Item> = vec![];
-    let mut items: Vec<syn::Item> = items
+    items = consume_derives(items, &metadata);
+    items = consume_impls(items, &metadata);
+    module.content = Some((braces, items));
+
+    let result = quote! {
+        #attrs
+        #[ink_lang::contract]
+        #module
+    };
+    result.into()
+}
+
+fn consume_traits(items: Vec<syn::Item>) -> Vec<syn::Item> {
+    items
+        .into_iter()
+        .filter_map(|mut item| {
+            if let Item::Trait(item_trait) = &mut item {
+                if is_attr(&item_trait.attrs, "storage_trait") {
+                    let attrs = item_trait.attrs.clone()
+                        .into_iter()
+                        .filter_map(|attr|
+                            if is_attr(&vec![attr.clone()],"storage_trait") {
+                                None
+                            } else {
+                                Some(attr)
+                            });
+                    item_trait.attrs.clear();
+
+                    let attrs = quote! { #(#attrs)* };
+                    let stream = storage_trait::generate(
+                        attrs.into(), item_trait.to_token_stream().into());
+                    let new_trait_item = syn::parse::<syn::Item>(stream)
+                        .expect("Can't parse generated storage trait");
+                    return Some(new_trait_item)
+                } else if is_attr(&item_trait.attrs, "trait_definition") {
+                    let attrs = item_trait.attrs.clone()
+                        .into_iter()
+                        .filter_map(|attr|
+                            if is_attr(&vec![attr.clone()], "trait_definition") {
+                                None
+                            } else {
+                                Some(attr)
```

Comment:
the logic is identical for storage_trait and trait_definition?

Reply:
Yes and no, in both cases we need to extract attribute, but in one case - trait_definition, in another - storage_trait

Reply:
ok I see, no need to overenginner with extracting 🙃 it's all good